vivarium/framework/components.py

```diff
@@ -1,63 +1,296 @@
+                    # it is idempotent so there's no effect.
+                    config.read_dict(component.configuration_defaults, layer='component_configs', source=component)
+
+                if hasattr(component, 'setup'):
+                    sub_components = component.setup(builder)
+                    done.append(component)
+                    if sub_components:
+                        components.extend(sub_components)
+                        self.components.extend(sub_components)
+
+
+def _extract_component_list(component_config: Mapping[str, Union[str, Mapping]]) -> Sequence[str]:
+    """Extract component descriptions from the hierarchical package/module groupings in the config file.
+
+    Parameters
+    ----------
+    component_config
+       The configuration to read from
+    """
+
+    def _process_level(level, prefix):
         component_list = []
-        for c in level:
-            if isinstance(c, dict):
-                for k, v in c.items():
-                    component_list.extend(process_level(v, prefix + [k]))
+        for child in level:
+            if isinstance(child, dict):
+                for path_suffix, sub_level in child.items():
+                    component_list.extend(_process_level(sub_level, prefix + [path_suffix]))
             else:
-                component_list.append('.'.join(prefix + [c]))
+                component_list.append('.'.join(prefix + [child]))
         return component_list
 
-    return process_level(component_config['components'], [])
+    return _process_level(component_config, [])
+
+
+def _component_ast_to_path(component: ast.AST) -> str:
+    """Convert the AST representing a component into a string which can be imported.
+
+    Parameters
+    ----------
+    component:
+        The node representing the component
+    """
+
+    if isinstance(component, ast.Name):
+        return component.id
+    path = []
+    current = component
+    while isinstance(current, ast.Attribute):
+        path.insert(0, current.attr)
+        current = current.value
+    path.insert(0, current.id)
+    return '.'.join(path)
+
+
+def _parse_component(desc: str, constructors: Mapping[str, Callable]) -> Tuple[str, Sequence]:
+    """Parse a component definition in a subset of python syntax and return an importable
+    path to the specified component along with the arguments it should receive when invoked.
+    If the definition is not parsable a ParsingError is raised.
+
+    If the component's arguments are not literals they are looked up in the constructors mapping
+    and if a compatible constructor is present it will be called. If no matching constructor is
+    available a ParsingError is raised.
+
+    Parameters
+    ----------
+    desc
+        The component definition
+    constructors
+        Dictionary of callables for creating argument objects
+    """
+
+    component, *args = ast.iter_child_nodes(list(ast.iter_child_nodes(
```

Comment:
This is maybe useful to have as a standalone function.  It needs an intelligible name and you also use it a few times in the tests.